File: core/agents/minimax.py
# Agent for Minimax Decisions
# Heuristic from "https://brilliant.org/discussions/thread/artificial-intelligence-for-chain-reaction/"

# -------- HEURISTIC FUNCTION --------
# 1. If player wins, score is 10000, else if the enemy wins, score is -10000.
# 2. For friendly orb, add 1 to the value.
# 3. For friendly orb, subtract 5 - critical mass of orb for every surrounding enemy critical cell.
# 4. For friendly safe orb, add 2 (edge cell) or 3 (corner cell).
# 5. For friendly critical safe orb, add 2 to the value.
# 6. For friendly critical non-orphan orb, add twice the number of cells to the score.
# ------------------------------------
# "friendly" : orb belonging to player
# "critical" : orb one step away from explosion
# "safe"     : orb with no adjacent critical enemy cells
# "orphan"   : orb with no adjacent critical friendly cells


import random
import core.engine as engine


# ------------ UTILITIES -------------
def moves_list(board, player) -> list:
    """ Get list of all valid moves for player on board """
    psign = -1 if player else 1
    return [i for i, orb in enumerate(board) if orb * psign >= 0]


def board_score(board, player) -> int:
    """ Calculate board score in favor of player """
    # setup
    size = len(board)
    psign = -1 if player else 1
    ntable = engine.NTABLE
    total_score = 0

    # cache tables for quick lookups
    en_crit = [board[x] * psign == 1 - len(ntable[x]) for x in range(size)]
    fr_crit = [board[x] * psign == len(ntable[x]) - 1 for x in range(size)]

    for idx in range(size):
        # multiplying psign makes player territories positive
        plr_orbs = board[idx] * psign

        # player territory
        if plr_orbs > 0:
            # assign to local variable
            neighbrs = ntable[idx]
            is_critc = fr_crit[idx]
            maxcp = len(neighbrs)

            # number of surrounding critical enemies and friends
            crit_enemies = sum([en_crit[nid] for nid in neighbrs])
            crit_friends = sum([fr_crit[nid] for nid in neighbrs])

            # RULE 2 and RULE 3
            total_score += plr_orbs
            total_score -= (5 - maxcp) * crit_enemies

            # RULE 4 and RULE 5
            if not crit_enemies:
                total_score += 3 if maxcp == 2 else 0
                total_score += 2 if maxcp == 3 else 0
                total_score += 2 if is_critc else 0

            # RULE 6
            if is_critc and crit_friends:
                total_score += 2

    return total_score


def minimizer(board, player, alpha, beta, depth) -> int:
    """ Minimizing Tree Search Function """

    # setup
    interact = engine.interact_view
    enemy = 1 - player
    score = 10000

    # searching all nodes
    for move in moves_list(board, enemy):

        # prune immediately if game over
        cboard, gmovr = interact(board, move, enemy)
        if gmovr:
            return -10000

        # get child score
        if depth:
            cscore = maximizer(cboard, player, alpha, beta, depth)
        else:
            cscore = board_score(cboard, player)

        # update
        score = min(score, cscore)
        beta = min(beta, score)

        # alpha-beta pruning
        if alpha >= beta:
            break

    return score


def maximizer(board, player, alpha, beta, depth) -> int:
    """ Maximizing Tree Search Function """

    # setup
    interact = engine.interact_view
    score = -10000

    # searching all nodes
    for move in moves_list(board, player):

        # prune immediately if game over
        cboard, gmovr = interact(board, move, player)
        if gmovr:
            return 10000

        # update score and beta
        score = max(score, minimizer(cboard, player, alpha, beta, depth - 1))
        alpha = max(alpha, score)

        # alpha-beta pruning
        if alpha >= beta:
            break

    return score


def move_chooser(board: list, player: int, depth=1, samples=3) -> int:
    """
    Get weighted random choice of best moves for the player
    -------------------------------------------------------
    If there is an immediate winning move, always return it
    board   : state of game where the choice has to be made
    player  : player id corresponding to the deciding agent
    depth   : every increment adds both max and mini levels
    samples : the top n moves which will be chosen randomly
    """

    # setup
    interct = engine.interact_view
    movlist = moves_list(board, player)
    heatmap = [0] * len(movlist)
    alpha = -10000

    # store values for sorting
    for i, move in enumerate(movlist):

        # return winning move immediately
        nboard, gmovr = interct(board, move, player)
        if gmovr:
            return move

        # store score and alpha
        score = minimizer(nboard, player, alpha, 10000, depth - 1)
        alpha = max(alpha, score)
        heatmap[i] = (move, score)

    # get random move with scores as weights
    # if score difference is high, probability of selection is low
    heatmap.sort(key=lambda x: x[1], reverse=True)
    bestmoves = [i[0] for i in heatmap[:samples] if i[1] > 0]
    bestscore = [i[1] for i in heatmap[:samples] if i[1] > 0]

    # return random choice if more than one score is positive
    # otherwise, return the move with highest score
    if len(bestmoves) > 1:
        return random.choices(bestmoves, bestscore)[0]
    else:
        return heatmap[0][0]

File: csource/src/minimax.c
#include "chain/engine.h"
#include "chain/minimax.h"


/* Critical Mass Lookup Table */
static char NTABLE [9 * 6] = {
    2, 3, 3, 3, 3, 2,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    2, 3, 3, 3, 3, 2,
};


/* Heuristic Evaluation Functions (in favor of player) */
static int WIN_SCORE = +10000;
static int LOS_SCORE = -10000;

static int
minimax__evaluation_score ( int  *board,
                            int   player )
{
    int psign = player ? -1 : 1;
    int score = 0;

    int c_frn_arr[54];
    int c_enm_arr[54];

    /* critical friends and enemy table */
    for (int i = 0; i < 54; ++i)
    {
        c_frn_arr[i] = ((board[i] * psign) == (NTABLE[i] - 1));
        c_enm_arr[i] = ((board[i] * psign) == (1 - NTABLE[i]));
    }

    for (int i = 0; i < 54; ++i)
    {
        int plr_orbs = board[i] * psign;

        /* skip enemy and empty territory */
        if (plr_orbs <= 0)
            continue;

        /* Critical friends and enemies */
        int crit_frn = 0;
        int crit_enm = 0;

        /* count surrounding enemies and friends */
        int i_y = i / 6;
        int i_x = i % 6;

        if (i_y > 0)
        {
            if (c_frn_arr[i - 6]) ++crit_frn;
            if (c_enm_arr[i - 6]) ++crit_enm;
        }
        if (i_y < 8)
        {
            if (c_frn_arr[i + 6]) ++crit_frn;
            if (c_enm_arr[i + 6]) ++crit_enm;
        }
        if (i_x > 0)
        {
            if (c_frn_arr[i - 1]) ++crit_frn;
            if (c_enm_arr[i - 1]) ++crit_enm;
        }
        if (i_x < 5)
        {
            if (c_frn_arr[i + 1]) ++crit_frn;
            if (c_enm_arr[i + 1]) ++crit_enm;
        }

        score += plr_orbs;
        score -= crit_enm * (5 - NTABLE[i]);

        if (crit_enm == 0)
        {
            if (NTABLE[i] == 2)             score += 3;
            if (NTABLE[i] == 3)             score += 2;
            if (NTABLE[i] == plr_orbs + 1)  score += 2;
        }

        if ((NTABLE[i] == plr_orbs + 1) && (crit_frn > 0))
            score += 2;
    }

    return score;
}


/* Direct Evaluation Minimizer Level */
int
minimax__score_minimizer  ( int  *board,
                            int   player,
                            int   alpha,
                            int   beta )
{
    /* Assume worst case score and improve */
    int score = WIN_SCORE;
    int enemy = 1 - player;
    int esign = enemy ? -1 : 1;
    int new_board[54];

    /* no more depth to explore */
    for (int i = 0; i < 54; ++i)
    {
        /* skip invalid moves (of enemy) */
        if (board[i] * esign < 0)
            continue;

        /* Interact with environment (enemy) */
        /* Node search is done if game over */
        if (engine__interact(board, new_board, i, enemy))
            return LOS_SCORE;
        
        /* Get recursive score and minimize score and beta */
        int child_score = minimax__evaluation_score(new_board, player);
        score = (child_score < score) ? child_score : score;
        beta  = (beta < score) ? beta : score;

        /* Node search is done if alpha >= beta */
        if (alpha >= beta)
            return score;
    }

    /* Return after search is completed */
    return score;
}


/* Minimax Minimizer Level (RECURSIVE) */
int
minimax__pruned_minimizer ( int  *board,
                            int   player,
                            int   alpha,
                            int   beta,
                            int   depth )
{
    /* Assume worst case score and improve */
    int score = WIN_SCORE;
    int enemy = 1 - player;
    int esign = enemy ? -1 : 1;
    int new_board[54];

    /* maximum depth reached => return min of scores instead */
    if (depth == 0)
        return minimax__score_minimizer(board, player, alpha, beta);

    /* more depth to explore */
    for (int i = 0; i < 54; ++i)
    {
        /* skip invalid moves (of enemy) */
        if (board[i] * esign < 0)
            continue;

        /* Interact with environment (enemy) */
        /* Node search is done if game over */
        if (engine__interact(board, new_board, i, enemy))
            return LOS_SCORE;
        
        /* Get recursive score and minimize score and beta */
        int child_score = minimax__pruned_maximizer(new_board, player, alpha, beta, depth);
        score = (child_score < score) ? child_score : score;
        beta  = (beta < score) ? beta : score;

        /* Node search is done if alpha >= beta */
        if (alpha >= beta)
            return score;
    }

    /* Return after search is completed */
    return score;
}


/* Minimax Maximizer Level (RECURSIVE) */
int
minimax__pruned_maximizer ( int  *board,
                            int   player,
                            int   alpha,
                            int   beta,
                            int   depth )
{
    /* Assume worst case score and improve */
    int score = LOS_SCORE;
    int psign = player ? -1 : 1;
    int new_board[54];

    /* more depth to explore */
    for (int i = 0; i < 54; ++i)
    {
        /* skip invalid moves */
        if (psign * board[i] < 0)
            continue;

        /* Node search is done if game over */
        if (engine__interact(board, new_board, i, player))
            return WIN_SCORE;
        
        /* Get recursive score and maximize score and alpha */
        int child_score = minimax__pruned_minimizer(new_board, player, alpha, beta, depth - 1);
        score = (child_score > score) ? child_score : score;
        alpha = (alpha > score) ? alpha : score;

        /* Node search is done if alpha >= beta */
        if (alpha >= beta)
            return score;
    }

    /* Return after search is completed */
    return score;
}